File: nodejs/base/api/lib/routing/defaults/routes/default-get-builder.ts
import { McmaResource, getTableName, McmaResourceType, Utils } from "@mcma/core";
import { DocumentDatabaseTableProvider } from "@mcma/data";
import { DefaultRouteHandlerConfigurator } from "../default-route-handler-configurator";
import { DefaultRouteBuilder } from "../default-route-builder";

export function defaultGetBuilder<T extends McmaResource>(
    dbTableProvider: DocumentDatabaseTableProvider,
    root: string
): DefaultRouteBuilder<T> {
    return new DefaultRouteBuilder<T>(
        "GET",
        root + "/{id}",
        new DefaultRouteHandlerConfigurator<T>((onStarted, onCompleted) =>
            async (requestContext) => {
                if (onStarted) {
                    const continueRequest = await onStarted(requestContext);
                    if (continueRequest !== undefined && !continueRequest) {
                        return;
                    }
                }

                const dbTable = await dbTableProvider.get(getTableName(requestContext));

                const resource = await dbTable.get<T>(requestContext.request.path);
                if (onCompleted) {
                    await onCompleted(requestContext, resource);
                }
                if (resource) {
                    requestContext.setResponseBody(resource);
                }
                else {
                    requestContext.setResponseResourceNotFound();
                }
            }
        )
    );
}

File: dotnet/base/Mcma.Api/Routing/Defaults/Routes/DefaultGetBuilder.cs
using System.Net.Http;
using Mcma;
using Mcma.Context;
using Mcma.Data;

namespace Mcma.Api.Routing.Defaults.Routes
{
    public static class DefaultGetBuilder
    {
        public static DefaultRouteBuilder<TResource> Get<TResource>(IDocumentDatabaseTableProvider dbTableProvider, string root) where TResource : McmaResource
            => 
            new DefaultRouteBuilder<TResource>(
                HttpMethod.Get,
                root + "/{id}",
                new DefaultRouteHandlerConfigurator<TResource>(
                    (onStarted, onCompleted) =>
                        async requestContext =>
                        {
                            // invoke the start handler, if any
                            if (onStarted != null)
                                if (!await onStarted(requestContext))
                                    return;

                            var table = await dbTableProvider.GetAsync(requestContext.TableName());

                            // get the resource from the database
                            var resource = await table.GetAsync<TResource>(requestContext.Request.Path);

                            // invoke the completion handler, if any
                            if (onCompleted != null)
                                await onCompleted(requestContext, resource);

                            // return the resource as json, if found; otherwise, this will return a 404
                            if (resource != null)
                                requestContext.SetResponseBody(resource);
                            else
                                requestContext.SetResponseResourceNotFound();
                        }));
    }
}